test/functional/lcow_uvm_test.go

```diff
@@ -0,0 +1,430 @@
+			},
+			wantArgs: []string{`root=/dev/sda`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/pmem`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+		{
+			name: "no SCSI single vPMEM VHD vmlinux",
+			optsFn: func(opts *uvm.OptionsLCOW) {
+				opts.SCSIControllerCount = 0
+				opts.VPMemDeviceCount = 1
+
+				opts.PreferredRootFSType = uvm.PreferredRootFSTypeVHD
+				opts.RootFSFile = uvm.VhdFile
+
+				opts.KernelDirect = true
+				opts.KernelFile = uvm.UncompressedKernelFile
+			},
+			wantArgs: []string{`root=/dev/pmem0`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/sda`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+		{
+			name: "SCSI no vPMEM VHD vmlinux",
+			optsFn: func(opts *uvm.OptionsLCOW) {
+				opts.SCSIControllerCount = 1
+				opts.VPMemDeviceCount = 0
+
+				opts.PreferredRootFSType = uvm.PreferredRootFSTypeVHD
+				opts.RootFSFile = uvm.VhdFile
+
+				opts.KernelDirect = true
+				opts.KernelFile = uvm.UncompressedKernelFile
+			},
+			wantArgs: []string{`root=/dev/sda`, `rootwait`, `init=/init`,
+				`8250_core.nr_uarts=0`, fmt.Sprintf(`nr_cpus=%d`, numCPU), `panic=-1`, `quiet`, `pci=off`},
+			notWantArgs:  []string{`initrd=`, `/dev/pmem`, `console=`},
+			notWantDmesg: []string{`initrd`, `initramfs`},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			opts := defaultLCOWOptions(ctx, t)
+			opts.ProcessorCount = numCPU
+			tc.optsFn(opts)
+
+			if opts.KernelDirect {
+				require.Build(t, 18286)
+			}
+
+			vm := testuvm.CreateAndStartLCOWFromOpts(ctx, t, opts)
+
+			// validate the kernel args were constructed as expected
+			ioArgs := testcmd.NewBufferedIO()
+			cmdArgs := testcmd.Create(ctx, t, vm, &specs.Process{Args: []string{"cat", "/proc/cmdline"}}, ioArgs)
+			testcmd.Start(ctx, t, cmdArgs)
+			testcmd.WaitExitCode(ctx, t, cmdArgs, 0)
+
+			ioArgs.TestStdOutContains(t, tc.wantArgs, tc.notWantArgs)
+
+			// some boot options (notably using initrd) need to validated by looking at dmesg logs
+			// dmesg will output the kernel command line as
+			//
+			// 	[    0.000000] Command line: <...>
+			//
+			// but its easier/safer to read the args directly from /proc/cmdline
+
+			ioDmesg := testcmd.NewBufferedIO()
+			cmdDmesg := testcmd.Create(ctx, t, vm, &specs.Process{Args: []string{"dmesg"}}, ioDmesg)
+			testcmd.Start(ctx, t, cmdDmesg)
+			testcmd.WaitExitCode(ctx, t, cmdDmesg, 0)
+
+			ioDmesg.TestStdOutContains(t, tc.wantDmesg, tc.notWantDmesg)
+		})
+	}
+}
+
+// TestLCOWUVM_Boot starts and terminates a utility VM  multiple times using different boot options.
+func TestLCOW_UVM_Boot(t *testing.T) {
+	require.Build(t, osversion.RS5)
+	requireFeatures(t, featureLCOW, featureUVM)
+
+	numIters := 3
```

Comment:
What is the purpose of doing 3 iterations of this? If it is to catch any concurrency issues, should we do 3 starts in 3 parallel goroutines?

Reply:
its in serial, and the original tests had it (https://github.com/microsoft/hcsshim/blob/575f7f83a743b64ea554eb6f41ea453704425641/test/functional/lcow_test.go#L170), so i just kept it